src/System.Linq/src/System/Linq/Set.cs

```diff
@@ -191,6 +191,38 @@ internal List<TElement> ToList()
         }
 
         /// <summary>
+        /// Converts this set to a <see cref="HashSet{T}"/>.
+        /// </summary>
+        /// <param name="comparer">The equality comparer for the new <see cref="HashSet{T}"/>.</param>
+        /// <returns>A <see cref="HashSet{T}"/> of the items in this set.</returns>
+        internal HashSet<TElement> ToHashSet(IEqualityComparer<TElement> comparer)
```

Comment:
@JonHanna Sorry, I saw your comment last week but forgot to reply.

> Once the test for the list-provider interface has been made there's no further runtime cost of optimising further, and I think it's clear enough what's happening so the cost to contributor comprehension isn't high either IMO.

Yes, but is it worth the extra code? Why optimize for a codepath when 1) the optimization isn't trivial, because it creates a new helper function (`Set.ToHashSet`) that has to be maintained, and 2) it won't be used because everyone can do `set = new HashSet<T>(first); set.UnionWith(second)` instead?